### ISSUE TITLE
Make round-tripping dates and datetimes consistent between CSV, JSON and NDJSON files

### ISSUE THREAD

**Issue by henryharbeck** [pola-rs/polars#19258]
### Description

TL;DR - try to get an immediate date or datetime with schema overrides when reading a CSV, JSON or NDJSON file is inconsistent. The table below shows results detailed below.

| data type | file type | stringified repr | unix repr
| --- | --- | --- | --- |
| date | CSV | works | raises
| date | JSON | null | works
| date | NDJSON | works | null
| datetime | CSV | works | raises
| datetime | JSON | works | null
| datetime | NDJSON | works | null

Of the types Polars does I/O with, CSV, JSON and NDJSON files do not have date types.

At the moment there are some differences with how dates and datetimes can be roundtripped via these file formats.

Starting with an example df
```python
df = (
    pl.select(d=pl.date(2024, 1, 1), dt=pl.datetime(2024, 1, 1))
    .select(
        "d",
        pl.col("d").dt.epoch(time_unit="d").alias("d_unix_days"),
        "dt",
        pl.col("d").dt.epoch(time_unit="us").alias("dt_unix_micros"),
    )
)
# shape: (1, 4)
# ┌────────────┬─────────────┬─────────────────────┬──────────────────┐
# │ d          ┆ d_unix_days ┆ dt                  ┆ dt_unix_micros   │
# │ ---        ┆ ---         ┆ ---                 ┆ ---              │
# │ date       ┆ i32         ┆ datetime[μs]        ┆ i64              │
# ╞════════════╪═════════════╪═════════════════════╪══════════════════╡
# │ 2024-01-01 ┆ 19723       ┆ 2024-01-01 00:00:00 ┆ 1704067200000000 │
# └────────────┴─────────────┴─────────────────────┴──────────────────┘
```

## CSV ##
In order to be read directly into a date/datetime type, it looks like:
- the stringified date works fine and the days since unix epoch (internal representation of date type) raises
- the stringified datetime works fine and the microseconds since unix epoch (default internal representation of datetime type) raises

```python
print(df.write_csv())
# d,d_unix_days,dt,dt_unix_micros
# 2024-01-01,19723,2024-01-01T00:00:00.000000,1704067200000000

pl.read_csv(
    df.write_csv().encode(),
    schema_overrides={
        "d": pl.Date, "d_unix_days": pl.Date,
        "dt": pl.Datetime, "dt_unix_micros": pl.Datetime,
    },
)
# ComputeError: could not parse `19723` as dtype `date` at column 'd_unix_days' (column number 2)
# OR if the `"d_unix_days": pl.Date` override is removed
# ComputeError: could not parse `1704067200000000` as dtype `datetime[μs]` at column 'dt_unix_micros' (column number 4)

pl.read_csv(
    df.write_csv().encode(),
    schema_overrides={"d": pl.Date, "dt": pl.Datetime},
)
# shape: (1, 4)
# ┌────────────┬─────────────┬─────────────────────┬──────────────────┐
# │ d          ┆ d_unix_days ┆ dt                  ┆ dt_unix_micros   │
# │ ---        ┆ ---         ┆ ---                 ┆ ---              │
# │ date       ┆ i64         ┆ datetime[μs]        ┆ i64              │
# ╞════════════╪═════════════╪═════════════════════╪══════════════════╡
# │ 2024-01-01 ┆ 19723       ┆ 2024-01-01 00:00:00 ┆ 1704067200000000 │
# └────────────┴─────────────┴─────────────────────┴──────────────────┘
```

## JSON ## 
In order to be read directly into a date/datetime type, it looks like:
- dates need to be roundtripped as the days since unix epoch (internal representation of date type), and the stringified version returns null. This is particularly puzzling, and is the only case out of all examples where the stringified version does not work
- the stringified datetime works fine and the microseconds since unix epoch (default internal representation of datetime type) returns null

```python
print(df.write_json())
# [{"d":"2024-01-01","d_unix_days":19723,"dt":"2024-01-01 00:00:00","dt_unix_micros":1704067200000000}]

pl.read_json(
    df.write_json().encode(),
    schema_overrides={
        "d": pl.Date, "d_unix_days": pl.Date,
        "dt": pl.Datetime, "dt_unix_micros": pl.Datetime,
    },
)
# shape: (1, 4)
# ┌──────┬─────────────┬─────────────────────┬────────────────┐
# │ d    ┆ d_unix_days ┆ dt                  ┆ dt_unix_micros │
# │ ---  ┆ ---         ┆ ---                 ┆ ---            │
# │ date ┆ date        ┆ datetime[μs]        ┆ datetime[μs]   │
# ╞══════╪═════════════╪═════════════════════╪════════════════╡
# │ null ┆ 2024-01-01  ┆ 2024-01-01 00:00:00 ┆ null           │
# └──────┴─────────────┴─────────────────────┴────────────────┘
```

## NDJSON ##
In order to be read directly into a date/datetime type, it looks like:
- the stringified date works fine and the days since unix epoch (internal representation of date type) returns null
- the stringified datetime works fine and the microseconds since unix epoch (default internal representation of datetime type) returns null

```python
print(df.write_ndjson())
# {"d":"2024-01-01","d_unix_days":19723,"dt":"2024-01-01 00:00:00","dt_unix_micros":1704067200000000}

pl.read_ndjson(
    df.write_ndjson().encode(),
    schema_overrides={
        "d": pl.Date, "d_unix_days": pl.Date,
        "dt": pl.Datetime, "dt_unix_micros": pl.Datetime,
    },
)
# shape: (1, 4)
# ┌────────────┬─────────────┬─────────────────────┬────────────────┐
# │ d          ┆ d_unix_days ┆ dt                  ┆ dt_unix_micros │
# │ ---        ┆ ---         ┆ ---                 ┆ ---            │
# │ date       ┆ date        ┆ datetime[μs]        ┆ datetime[μs]   │
# ╞════════════╪═════════════╪═════════════════════╪════════════════╡
# │ 2024-01-01 ┆ null        ┆ 2024-01-01 00:00:00 ┆ null           │
# └────────────┴─────────────┴─────────────────────┴────────────────┘
```
